test/framework/control_plane.go

```diff
@@ -69,81 +78,118 @@ func (input *ControlplaneClusterInput) ControlPlaneCluster() {
 	By("creating an InfrastructureCluster resource")
 	Expect(mgmtClient.Create(ctx, input.InfraCluster)).To(Succeed())
 
+	// This call happens in an eventually because of a race condition with the
+	// webhook server. If the latter isn't fully online then this call will
+	// fail.
 	By("creating a Cluster resource linked to the InfrastructureCluster resource")
 	Eventually(func() error {
-		err := mgmtClient.Create(ctx, input.Cluster)
-		if err != nil {
-			fmt.Println(err)
+		if err := mgmtClient.Create(ctx, input.Cluster); err != nil {
+			fmt.Printf("%+v\n", err)
+			return err
 		}
-		return err
-	}, input.CreateTimeout, 10*time.Second).Should(BeNil())
+		return nil
+	}, input.CreateTimeout, eventuallyInterval).Should(BeNil())
+
+	By("creating related resources")
+	for _, obj := range input.RelatedResources {
+		By(fmt.Sprintf("creating a/an %s resource", obj.GetObjectKind().GroupVersionKind()))
+		Eventually(func() error {
+			return mgmtClient.Create(ctx, obj)
+		}, input.CreateTimeout, eventuallyInterval).Should(BeNil())
+	}
 
-	// create all the machines at once
-	for _, node := range input.Nodes {
-		By("creating an InfrastructureMachine resource")
-		Expect(mgmtClient.Create(ctx, node.InfraMachine)).To(Succeed())
+	// expectedNumberOfNodes is the number of nodes that should be deployed to
+	// the cluster. This is the control plane nodes plus the number of replicas
+	// defined for a possible MachineDeployment.
+	expectedNumberOfNodes := 1
 
-		By("creating a bootstrap config")
-		Expect(mgmtClient.Create(ctx, node.BootstrapConfig)).To(Succeed())
+	By("creating the primary control plane node's InfrastructureMachine resource")
```

Comment:
Cluster API should be declarative. I'd consider Cluster API to have a bug if this is not the case. I realize the test is already breaking this assumption by creating the cluster first and then the control planes etc, but in an ideal world we'd be moving toward "create everything and then wait" and not towards creating a dependency graph of objects. 

However, if this is a requirement of CAPV, I'm fine adding a new, separate test that does require an order of object applications as it is possible that not all providers have the luxury of being completely declarative.

Reply:
It worked on CAPV. I think my original thinking was incorrect based on some issues I saw in-flight. I will change the PR to reflect this. Thanks!

Reply:
This should be resolved now, thanks!